File: C++/WorkoutPrinter/Month.cpp
#include "Month.h"
#include <iomanip>
#include <iostream>
#include <cmath>
#include <algorithm>
#include <iterator>

// default constructor
Month::Month(){}

// paremeterized constructor I
Month::Month(int month, int year){
	this->month = month;
	this->year = year;
	setNumDays(month, year);
	setFirstDay();
	generateDays(); 
}

// paremeterized constructor II
Month::Month(int month, int year, std::string prev_mg, std::string cur_mg, std::string next_mg){
	this->month = month;
	this->year = year;
	setNumDays(month, year);
	setFirstDay();
	generateBoesSplitDays(prev_mg, cur_mg, next_mg); 
}

// copy constructor
Month::Month(Month& new_month){
	setMonth(new_month.getMonth());
	setYear(new_month.getYear());
	setNumDays(month, year);
	setFirstDay();
	generateDays();
}

// destructor
Month::~Month(){
	for(Week* week : weeks){
		delete week;
	}
}

// implementation of setNumDays 
// sets the number of days for the inputted month and year
void Month::setNumDays(int month, int year){
	
	if(month == 2 && isLeapYear(year)){
		num_days = DAYS_IN_MONTH[month - 1] + 1;
	} else {
		num_days = DAYS_IN_MONTH[month - 1];
	}
}

// generating the months days *this does not add the muscle groups
void Month::generateDays(){
	int day_count = 1;
	int numWhileLoops = 0;
	
	
	while(day_count <= num_days){
		Week* newWeek = new Week(); 
		
		// populating new week
		for(int day = 0; day < 7; day++){
			if((numWhileLoops == 0 && day < first_day) || day_count > num_days){ // for other months days
				newWeek->addDay(new Day());
			} else { // for our months days
				newWeek->addDay(new Day(day_count));
				day_count++;
			}
		}
		weeks.push_back(newWeek);
		numWhileLoops++;
	}
}

// generatings the months days and the muscle group to train for that day
void Month::generateBoesSplitDays(std::string prev_mg, std::string cur_mg, std::string next_mg){
	int day_count = 1;
	int numWhileLoops = 0;
	std::string temp = "";

	while(day_count <= num_days){
		Week* newWeek = new Week();
		
		// populating new week
		for(int day = 0; day < 7; day++){
			if((numWhileLoops == 0 && day < first_day) || day_count > num_days){ // for other months days
				newWeek->addDay(new Day());
			} else { // for our months days
				if(day == 1 || day == 3 || day == 5) {
					newWeek->addDay(new Day(day_count, cur_mg));
					day_count++;
				} else if (day == 2){
					newWeek->addDay(new Day(day_count, next_mg));
					day_count++;
				} else if (day == 4){
					newWeek->addDay(new Day(day_count, prev_mg));
					day_count++;
				} else {
					newWeek->addDay(new Day(day_count));
					day_count++;
				}
			}
		}
		weeks.push_back(newWeek);
		numWhileLoops++;
		
		// preping muscle groups for the next week
		temp = cur_mg;
		cur_mg = next_mg;
		next_mg = prev_mg;
		prev_mg = temp;
	}	
}


// implementation of displayMonth
// prints the month
void  Month::displayMonth(){
	
	for(Week* week : weeks){
		week->displayWeek();
	}
}

// returns true if year is a leap year and false otherwise
bool Month::isLeapYear(int year){
	return (year % 4 == 0 && year % 100 != 0) || (year % 400 == 0);
}

void Month::setFirstDay(){
	
	int loc_month = month; // local copy of month
	int days_this_year = 0; // to hold number of days this year
	
	int starting_day = 4; // starting day of the week index from 2000
	
	// counting days already passed this year
	while(loc_month > 0){
		days_this_year += DAYS_IN_MONTH[loc_month];
		loc_month -= 1;
	}
	
	// we begin with the year 2000 and add up the days that have passed since then
	for(int i = 2000; i < year; i++){
		if(isLeapYear(i)){
			starting_day += 366;
		} else {
			starting_day += 365;
		}
	}
	
	first_day = (starting_day + days_this_year) % 7;	
}

// getters
int Month::getMonth(){
	return month;
}

int Month::getYear(){
	return year;
}

std::string Month::getMessage(){
	return message;
}

int Month::getFirstDayIndex(){
	return first_day;
}

std::vector<Week*> Month::getWeeks(){
	return weeks;
}

// setters
void Month::setYear(int new_year){
	this->year = new_year;
}

void Month::setMonth(int new_month){
	this->month = new_month;
}

File: Java/WorkoutPrinter/src/Month.java
public class Month{

    private final int[] DAYS_IN_MONTH = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
    private final int CELL_WIDTH = 15;
    private int month;
	private int year;
	private int numDays;
	private int firstDay;
    private Day[][] days;

    public Month(int month, int year){
        this.month = month;
        this.year = year;
        days = new Day[6][7];
        setNumDays(month, year);
        setFirstDay();
        generateDays("Back and Biceps", "Chest and Triceps", "Legs");
    }

	private boolean isLeapYear(int year){
        return (year % 4 == 0 && year % 100 != 0) || (year % 400 == 0);
    }

	private void setFirstDay(){
        int loc_month = month; // local copy of month
        int daysThisYear = 0; // to hold number of days this year
        
        int startingDay = 6; // starting day from 1800
        
        // counting days already passed this year
        while(loc_month > 0){
            daysThisYear += DAYS_IN_MONTH[loc_month - 1];
            loc_month -= 1;
        }
        
        for(int i = 1800; i < year; i++){
            if(isLeapYear(year)){
                startingDay += 366;
            } else {
                startingDay += 365;
            }
        }
        
        firstDay = (startingDay + daysThisYear) % 7;	
    }

    public void setNumDays(int month, int year){
        if(month == 2 && isLeapYear(year)){
            numDays = DAYS_IN_MONTH[month - 1] + 1;
        } else {
            numDays = DAYS_IN_MONTH[month - 1];
        }
    }

	public void displayMonth(){
        for(int row = 0; row < 6; ++row){ // weeks
            for(int col = 0; col < 7; col++){ // day of the week
                if(days[row][col].getDayNum() != 0 || row+col==11 || col == 0){
                    if(col == 6){
                        if(days[row][col].getDayNum() == 0){
                            System.out.printf("%21s", "|\n"); 
                        } else {
                            System.out.printf("%21s", Integer.toString(days[row][col].getDayNum()) + "|\n");
                        }
                    } else if (col == 0) {
                        System.out.print("|");

                        System.out.printf("%20s", Integer.toString(days[row][col].getDayNum()) + "|");

                    } else {
                        System.out.printf("%20s", Integer.toString(days[row][col].getDayNum()) + "|");
                    }
                } else {
                    if(col == 6) {
                        System.out.printf("%20s", "|");
                        System.out.println();
                    } else {
                        System.out.printf("%20s", "|");
                    }
                }
            }
            //std::cout << '\n';
    
            printSquares();
            
        }
    }

    void printSquares(){
        for(int r = 0; r < 6; r++){ // increase r's limit to print longer cells
            for(int c = 0; c < 7; c++){

                if(c == 0){
                    System.out.print("|");
                }

                if(r == 0){
                    System.out.printf("%20s", days[r][c].getMuscleGroup() + "|");
                } else {
                    System.out.printf("%20s", "|");
                }             
                
            }
            
            System.out.println();
            
        }
        
        System.out.println(Calendar.HORIZONTAL_LINE);
    }

	public int getMonth(){
        return month;
    }

	public int getYear(){
        return year;
    }

    private void generateDays(){
        int dayCount = 1;
	
        for(int row = 0; row < 6; row++){ // weeks
            for(int col = 0; col < 7; col++){ // day of the week
                if(row == 0 && col < firstDay){ // cells before the first day
                    days[row][col] = new Day(); // make blank day
                } else if (dayCount > numDays){ // cells past the month's last day
                    days[row][col] = new Day(); // make blank day
                } else {
                    days[row][col] = new Day(dayCount); // make day initialized with dayCount
                    dayCount++; // increment dayCount
                }
            }
        }
    }

    private void generateDays(String curWeekMuscleGroup, String nextWeekMuscleGroup, String prevWeekMuscleGroup){
        int dayCount = 1;
        String temp = "";
	
        for(int row = 0; row < 6; row++){ // weeks
            for(int col = 0; col < 7; col++){ // day of the week
                if(row == 0 && col < firstDay){ // cells before the first day
                    days[row][col] = new Day(); // make blank day
                } else if (dayCount > numDays){ // cells past the month's last day
                    days[row][col] = new Day(); // make blank day
                } else { // setting actual month days
                    days[row][col] = new Day(dayCount); // make day initialized with dayCount
                    switch(col){ // setting muscle group based on weekday
                        case 1: // Monday
                            days[row][col].setMuscleGroup(curWeekMuscleGroup);
                        case 2: // Teusday
                            days[row][col].setMuscleGroup(nextWeekMuscleGroup);
                        case 3: // Wednesday
                            days[row][col].setMuscleGroup(curWeekMuscleGroup);
                        case 4: // Thurseday
                            days[row][col].setMuscleGroup(prevWeekMuscleGroup);
                        case 5: //Friday
                            days[row][col].setMuscleGroup(curWeekMuscleGroup);
                        default:
                            break;
                    }

                    //days[row][col].setMuscleGroup(curWeekMuscleGroup);
                    dayCount++; // increment dayCount
                }
            }
            temp = curWeekMuscleGroup;
            curWeekMuscleGroup = nextWeekMuscleGroup;
            nextWeekMuscleGroup = prevWeekMuscleGroup;
            prevWeekMuscleGroup = temp;
        }
    }

}